cmd/juju/caas/update.go

```diff
@@ -0,0 +1,406 @@
+		}
+		return builtinCloud, "", credential, credentialName, nil
+	}
+
+	// Not a built-in cloud so extract the details from kubeconfig.
+	rdr, closer, clusterName, err := c.getK8sConfigReader(ctx)
+	if err != nil {
+		return fail(errors.Trace(err))
+	}
+	defer closer()
+	config, err := c.getK8sCloudParams(clusterName)
+	if err != nil {
+		return fail(errors.Trace(err))
+	}
+
+	newCloud, newCredential, err := provider.CloudFromKubeConfig(rdr, *config)
+	if err != nil {
+		return fail(errors.Trace(err))
+	}
+	credentialName = c.caasName
+	newCredential, err = ensureCredentialUID(credentialName, config.CredentialUID, newCredential)
+	if err != nil {
+		return fail(errors.Trace(err))
+	}
+	return newCloud, clusterName, &newCredential, credentialName, nil
+}
+
+// Run is defined on the Command interface.
+func (c *UpdateCAASCommand) Run(ctx *cmd.Context) (err error) {
+	cloudToUpdate, clusterName, credToUpdate, credentialName, err := c.getCloudDetails(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// We need to have c.ControllerName after this, so this is the latest time to
+	// prompt user for client and controller options.
+	// We need to do this later then other commands since
+	// piping is done regularly with update-k8s.
+	if err := c.MaybePrompt(ctx, fmt.Sprintf("update k8s cloud %v to", c.caasName)); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Check the cluster only if we have a credential to use.
+	if credToUpdate != nil {
+		broker, err := c.brokerGetter(cloudToUpdate, *credToUpdate)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if _, err := broker.GetClusterMetadata(""); err != nil {
+			return errors.Annotate(err, "unable to update k8s cluster because the cluster is not accessible")
+		}
+	}
+
+	if cloudToUpdate.HostCloudRegion != "" {
+		cloudToUpdate.HostCloudRegion, err = c.validateCloudRegion(ctx, cloudToUpdate.HostCloudRegion)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	// By this stage, we know if cloud name/type and/or region input is needed from the user.
+	// If we could not detect it, check what was provided.
+	if err := checkCloudRegion(c.givenHostCloudRegion, cloudToUpdate.HostCloudRegion); err != nil {
+		return errors.Trace(err)
+	}
+	if cloudToUpdate.HostCloudRegion == "" {
+		cloudToUpdate.HostCloudRegion = caas.K8sCloudOther
+	}
+	if c.Client {
+		if err := updateCloudOnLocal(c.cloudMetadataStore, cloudToUpdate); err != nil {
+			return errors.Trace(err)
+		}
+		if credToUpdate != nil {
+			if err := c.addCredentialToLocal(c.credentialStoreAPI, c.caasName, *credToUpdate, credentialName); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+
+	if clusterName == "" && cloudToUpdate.HostCloudRegion != caas.K8sCloudOther {
+		clusterName = cloudToUpdate.HostCloudRegion
+	}
+	if clusterName != "" {
+		clusterName = fmt.Sprintf("%q ", clusterName)
+	}
+	successMsg := fmt.Sprintf("k8s substrate %sknown to Juju as cloud %q updated.", clusterName, c.caasName)
```

Comment:
`%s known`

Reply:
The substituted value for clusterName has a " " at the end (it may be empty).